pythonfmu/fmi2slave.py

```diff
@@ -251,12 +252,17 @@ def set_string(self, vrs: List[int], values: List[str]):
                     f"Variable with valueReference={vr} is not of type String!"
                 )
 
-    def _get_fmu_state(self) -> Dict[str, Any]:
+    def store_custom_state(self, state: Dict[str, Any]):
+        pass
+
+    def _get_fmu_state(self) -> bytes:
```

Comment:
I would rather keep the `get/set_fmu_state` as proposed in #53 and add `get/set_serialized_state` that carry the json (de)serialization. This will keep things closer to FMI and allow customization with different state container object and/or serialization logic.

Reply:
> This will keep things closer to FMI 

How do you mean? get/set state and serialize/deserialize contains the exact same information in FMI. Only that the latter can be saved to file.

Reply:
`fmi2SerializeFMUstate` returns a pointer to the same object as you would get with `fmi2GetState`. So you can't distinguish between those two calls.

Reply:
It looks to me that the standard separate the serialization from the state handlers as it may not be straightforward. 

This is for example true for the use case, you proposed in which non standard type could be added to the state (and therefore probably non-jsonable type neither).

To provide an example using numpy array of dimension 0:

```python
In [1]: import numpy as np

In [2]: a = np.array(2)

In [3]: int(a)
Out[3]: 2

In [4]: import json

In [5]: b = {'a': a}

In [6]: json.dumps(b)
---------------------------------------------------------------------------
TypeError: Object of type ndarray is not JSON serializable
```

So if you split the serialization from the getter, the get state will work with numpy.array. But serialization through JSON won't work. And this can be specified when building the FMU through the options: `canGetAndSetFMUstate=true` and `canSerializeFMUstate=false`.

But with the current proposal, this force the user to use only JSONable type even to only support get/set FMU state or to overwrite the get/set FMU state methods.

So although this requires adding new Python methods, I think it is a more flexible approach.

Reply:
The thing is that get/set state and serialize/deserialize state is basically the same function in FMI. Serialize state takes the pointer given to it by getState and converts it bytes and deserialize turns it back into the same pointer you'd get calling getState at that point in time. What I'm saying is that these functions are intertwined. We can't have special functions for the two cases. However, it's up to the user to say to the builder that ok, my FMU can handle serialization as well.

Reply:
Ok, ok. What you are saying is that this could potentially break code that only need get/set state? Right, I may see your point. The alternative to this how I see it is to pass the dict object back into python for additional work in the serialize/deserialize case.

Reply:
Yes this is what I mean - sorry thought are sometimes hard to write down :wink: